core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala

```diff
@@ -446,6 +448,94 @@ trait DeltaSQLConfBase {
       .booleanConf
       .createWithDefault(false)
 
+  final object MergeMaterializeSource {
+    // See value explanations in the doc below.
+    final val NONE = "none"
+    final val ALL = "all"
+    final val AUTO = "auto"
+
+    final val list = Set(NONE, ALL, AUTO)
+  }
+
+  val MERGE_MATERIALIZE_SOURCE =
+    buildConf("merge.materializeSource")
+      .internal()
+      .doc("When to materializes source plan during MERGE execution. " +
+        "The value 'none' means source will never be materialized. " +
+        "The value 'all' means source will always be materialized. " +
+        "The value 'auto' means sources will not be materialized when they are certain to be " +
```

Comment:
i am confused. source will be not materialized when they are deterministic? or when they are not deterministic?

Reply:
nvm.. i got it. i had it backwards in my head.